File: Sources/lambda/Utils/ResponseBuilder.swift
import AWSLambdaEvents
import Foundation
import HTTPTypes

enum ResponseBuilder {
    static let defaultCORSHeaders: [String: String] = [
        "content-type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET,OPTIONS,POST,PUT,DELETE"
    ]

    static func createResponse<T: Encodable>(
        body: T,
        includeCacheControl: Bool = true,
        revalidateSeconds: Int = 3
    ) throws -> APIGatewayV2Response {
        let encoder = JSONEncoder()
        var headers = defaultCORSHeaders

        if includeCacheControl {
            headers["Cache-Control"] =
                "max-age=\(revalidateSeconds), s-maxage=\(revalidateSeconds), stale-while-revalidate=\(revalidateSeconds), stale-if-error=\(revalidateSeconds)"
        } else {
            headers["Cache-Control"] = "no-cache"
        }

        let bodyData = try encoder.encode(body)
        let bodyString = String(data: bodyData, encoding: .utf8) ?? "{}"

        return APIGatewayV2Response(
            statusCode: .ok,
            headers: headers,
            body: bodyString
        )
    }

    static func errorResponse(statusCode: HTTPResponse.Status, message: String) -> APIGatewayV2Response {
        let body = (try? JSONEncoder().encode(["error": message])).flatMap { String(data: $0, encoding: .utf8) } ?? #"{"error":"Unknown error"}"#
        return APIGatewayV2Response(
            statusCode: statusCode,
            headers: defaultCORSHeaders,
            body: body
        )
    }
}

File: src/Lho.Lambda/Utils/ResponseBuilder.cs
using System.Text.Json;
using Amazon.Lambda.APIGatewayEvents;
using Lho.Lambda.Models;
using Lho.Lambda.Serialization;

namespace Lho.Lambda.Utils;

public static class ResponseBuilder
{
  private static readonly JsonSerializerOptions JsonOptions = new()
  {
    PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
    TypeInfoResolver = ApiResponseJsonContext.Default
  };

  private static readonly Dictionary<string, string> DefaultCorsHeaders = new()
  {
    ["content-type"] = "application/json",
    ["Access-Control-Allow-Origin"] = "*",
    ["Access-Control-Allow-Methods"] = "GET,HEAD,OPTIONS,POST,PUT,DELETE"
  };

  public static APIGatewayHttpApiV2ProxyResponse CreateResponse<T>(
    T body,
    bool includeCacheControl = true,
    int revalidateSeconds = 3)
  {
    var headers = new Dictionary<string, string>(DefaultCorsHeaders)
    {
      ["Cache-Control"] = includeCacheControl
        ? $"max-age={revalidateSeconds}, s-maxage={revalidateSeconds}, stale-while-revalidate={revalidateSeconds}, stale-if-error={revalidateSeconds}"
        : "no-cache"
    };

    return new APIGatewayHttpApiV2ProxyResponse
    {
      StatusCode = 200,
      Headers = headers,
      Body = JsonSerializer.Serialize(body, JsonOptions)
    };
  }

  public static APIGatewayHttpApiV2ProxyResponse ErrorResponse(int statusCode, string message)
  {
    return new APIGatewayHttpApiV2ProxyResponse
    {
      StatusCode = statusCode,
      Headers = new Dictionary<string, string>(DefaultCorsHeaders),
      Body = JsonSerializer.Serialize(new ErrorResponseBody(message), JsonOptions)
    };
  }
}
